src/sidebar/services/auth.ts

```diff
@@ -0,0 +1,335 @@
+   * Fetch the last-saved access/refresh tokens for `authority` from local
+   * storage.
+   */
+  private _loadToken() {
+    const token = this._localStorage.getObject(this._storageKey());
+
+    if (
+      !token ||
+      typeof token.accessToken !== 'string' ||
+      typeof token.refreshToken !== 'string' ||
+      typeof token.expiresAt !== 'number'
+    ) {
+      return null;
+    }
+
+    return {
+      accessToken: token.accessToken,
+      refreshToken: token.refreshToken,
+      expiresAt: token.expiresAt,
+    };
+  }
+
+  /**
+   * Return a cached OAuthClient if available, or configure one and return it.
+   */
+  async _oauthClient() {
+    if (this._client) {
+      return this._client;
+    }
+
+    const links = await this._apiRoutes.links();
+    this._client = new OAuthClient({
+      clientId: this._settings.oauthClientId,
+      authorizationEndpoint: links['oauth.authorize'],
+      revokeEndpoint: links['oauth.revoke'],
+      tokenEndpoint: resolve('token', this._settings.apiUrl),
+    });
+    return this._client;
+  }
+
+  /**
+   * Exchange the authorization code retrieved from login popup for a new
+   * access token.
+   */
+  private async _exchangeAuthCodeForToken(code: string) {
+    const client = await this._oauthClient();
+    const tokenInfo = await client.exchangeAuthCode(code);
+    this._saveToken(tokenInfo);
+    return tokenInfo;
+  }
+
+  /**
+   * Attempt to exchange a grant token for an access token.
+   *
+   * Publishers or other embedders can set a grant token for a user in service
+   * configuration.
+   */
+  private async _exchangeGrantToken(grantToken: string) {
+    if (grantToken) {
+      const client = await this._oauthClient();
+      // User is logged-in on the publisher's website.
+      // Exchange the grant token for a new access token.
+      return client.exchangeGrantToken(grantToken).catch(err => {
+        this._toastMessenger.error(
+          `Hypothesis login lost: You must reload the page to annotate.`,
+          {
+            autoDismiss: false,
+          }
+        );
+        throw err;
+      });
+    } else {
+      // Grant token is empty: user is anonymous on the publisher's website.
+      return null;
+    }
+  }
+
+  /**
+   * Retrieve any grant token set by service configuration
+   */
+  private _getGrantToken(): string | undefined {
+    return serviceConfig(this._settings)?.grantToken;
+  }
+
+  /**
+   * Return the storage key used for storing access/refresh token data for a
+   * given annotation service.
+   */
+  private _storageKey() {
+    // Use a unique key per annotation service. Currently OAuth tokens are only
+    // persisted for the default annotation service. If in future we support
+    // logging into other services from the client, this function will need to
+    // take the API URL as an argument.
+    let apiDomain = new URL(this._settings.apiUrl).hostname;
+
+    // Percent-encode periods to avoid conflict with section delimeters.
```

Comment:
What does "section delimiters" refer to here? Does this mean "sections of the localStorage key"?

Reply:
Noooooo idea. Not my comment. Let me see if I can track it down.

Reply:
Turns out I know just the guy to ask: it was introduced in this PR https://github.com/hypothesis/client/pull/494

There's a fair amount of discussion in that PR thread and I'm not immediately seeing the provenance of the "section delimiters" notion. @robertknight Does this spur any memories?

Reply:
Ah, yeah. That guy. I'm pretty sure he was referring to "."s in localStorage key names.